qiskit/pulse/filters.py

```diff
@@ -61,6 +64,60 @@ def filter_instructions(
     return filter_schedule
 
 
+@filter_instructions.register
+def _(
+    sched_blk: ScheduleBlock,
+    filters: List[Callable],
+    negate: bool = False,
+    recurse_subroutines: bool = True,
+) -> ScheduleBlock:
+    """A filtering function that takes a schedule_block and returns a schedule_block consisting of
+    filtered instructions.
+
+    Args:
+        sched_blk: A pulse schedule_block to be filtered.
+        filters: List of callback functions that take an instruction and return boolean.
+        negate: Set `True` to accept an instruction if a filter function returns `False`.
+            Otherwise the instruction is accepted when the filter function returns `False`.
+        recurse_subroutines: Set `True` to individually filter instructions inside of a subroutine
+            defined by the :py:class:`~qiskit.pulse.instructions.Call` instruction.
+
+    Returns:
+        Filtered pulse schedule_block.
+    """
+    from qiskit.pulse.transforms import inline_subroutines
+
+    target_sched_blk = sched_blk
+    if recurse_subroutines:
+        target_sched_blk = inline_subroutines(target_sched_blk)
+
+    def apply_filters_to_insts_in_scheblk(blk: ScheduleBlock) -> ScheduleBlock:
+        blk_new = ScheduleBlock.initialize_from(blk)
+        for element in blk.blocks:
+            if isinstance(element, ScheduleBlock):
+                inner_blk = apply_filters_to_insts_in_scheblk(element)
```

Comment:
@nkanazawa1989 
Thank you very much for your comment and advice! 👋 

> The [implementation](https://github.com/Qiskit/qiskit-terra/blob/3ebef17322c7714aedaa71967dee74fd455a83a0/qiskit/pulse/transforms/canonicalization.py#L206-L209) of this function intentionally preserves nested schedule blocks because they may have different alignment context (we cannot expose blocks to outer context without breaking alignment). This may cause unexpected behavior, since block to be filtered could be nested ScheduleBlock rather than instructions. I think you need a function that recursively returns all blocks regardless of alignment.

I hope I correctly understand your message. In my understanding, the above code lines of the new implementation enables us to recursively filter all instructions even in a nested ScheduleBlock. 

Please let me give one example. 
Please let us take the following as an example of nested ScheduleBlocks:
```py
block = ScheduleBlock(alignment_context=pulse.transforms.AlignSequential())
block.append(Play(Constant(10, 0.1), DriveChannel(0)))
block.append(Delay(5, DriveChannel(0)))
internal_block = ScheduleBlock(name="internal_blk")
internal_block.append(Play(Gaussian(10, 0.2, 2), DriveChannel(1)))
block.append(Call(subroutine=internal_block))
block.append(Play(Constant(5, 0.1), DriveChannel(1)))
block.draw()
```
![image](https://user-images.githubusercontent.com/47435718/233825474-f13bd482-ac3d-4979-85f6-0cdf8f2f76d2.png)

When I filter `Play` instruction setting `check_subroutine=False`, 
```py
block_filterd = block.filter(instruction_types=[Play], check_subroutine=False)
block_filterd.draw()
```
then I get
![image](https://user-images.githubusercontent.com/47435718/233825488-a87366bf-f978-4625-9920-81b9ce4ff4ba.png)
As expected, the Gaussian pulse is filtered out because the `Call` instruction remains `Call`. 

When I do the same setting `check_subroutine=True`, 
```py
block_filterd = block.filter(instruction_types=[Play], check_subroutine=True)
block_filterd.draw()
```
then I get
![image](https://user-images.githubusercontent.com/47435718/233825770-e8ce8de7-81bc-437c-9ac7-92870ed36fd6.png)
As this is also expected, the Gaussian pulse in the internal ScheduleBlock is correctly picked up because the `Call` instruction was transformed to a ShceduleBlock. So, I hope this result shows that the new implementation correctly filters instructions in nested ScheduleBlocks with the `check_subroutine=True` option. 

I would be grateful if you could give me a comment back. 🙇

Reply:
Thanks. This is outside the scope of your PR, but wrapping `ScheduleBlock` with `Call` doesn't have any benefit and we plan to deprecate it. See https://github.com/Qiskit/qiskit-terra/issues/10103 for more details. The PR looks good anyways.